OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs

```diff
@@ -14,28 +14,79 @@
 using System.Drawing;
 using System.Linq;
 using OpenRA.Primitives;
+using OpenRA.Support;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Traits.Radar
 {
-	public class AppearsOnRadarInfo : ITraitInfo
+	public class AppearsOnRadarInfo : IObservesVariablesInfo, IRulesetLoaded
 	{
+		static readonly IEnumerable<string> NoVariables = Enumerable.Empty<string>();
+
 		public readonly bool UseLocation = false;
 
+		[Desc("Boolean condition for radar visibility. Defaults to true.")]
+		public readonly BooleanExpression ShowOnCondition = null;
+
+		[Desc("Boolean condition for enemy radar visibility. Defaults to true.",
+			"ShowOnCondition must be true (or default).")]
+		public readonly BooleanExpression ShowEnemyOnCondition = null;
+
+		public IEnumerable<string> VariablesUsedForShowing { get { return variablesUsedForShowing; } }
+
+		IEnumerable<string> variablesUsedForShowing;
+
 		public object Create(ActorInitializer init) { return new AppearsOnRadar(this); }
+
+		public void RulesetLoaded(Ruleset rules, ActorInfo info)
+		{
+			if (ShowOnCondition != null)
+			{
+				if (ShowEnemyOnCondition != null)
+					variablesUsedForShowing = new List<string>(ShowOnCondition.Variables.Concat(ShowEnemyOnCondition.Variables).Distinct());
+				else
+					variablesUsedForShowing = ShowOnCondition.Variables;
+			}
+			else
+			{
+				if (ShowEnemyOnCondition != null)
+					variablesUsedForShowing = ShowEnemyOnCondition.Variables;
+				else
+					variablesUsedForShowing = NoVariables;
+			}
+		}
 	}
 
-	public class AppearsOnRadar : IRadarSignature, INotifyCreated
+	public class AppearsOnRadar : IObservesVariables, IRadarSignature, INotifyCreated
 	{
+		static readonly IEnumerable<Pair<CPos, Color>> NoCells = Enumerable.Empty<Pair<CPos, Color>>();
 		readonly AppearsOnRadarInfo info;
 		IRadarColorModifier modifier;
 
 		Color currentColor = Color.Transparent;
 		Func<Pair<CPos, SubCell>, Pair<CPos, Color>> cellToSignature;
+		Stance visibleStances = Stance.Ally | Stance.Enemy;
```

Comment:
There are quite a few maps that place neutral base structures or units on the map for decoration or to be captured.